### PR TITLE
Misc refactors and documentation

### PR DESCRIPTION
**Pull request by smithjessk**
*not done yet*

just opening for initial thoughts @neelnanda-io :)

### CODE REVIEW

File: easy_transformer/EasyTransformer.py [hunk 1]
```diff
@@ -126,6 +106,39 @@ def __init__(
...
+            provided, it is inferred from cfg.tokenizer_name or initialized to None.
+            If None, then the model cannot be passed strings, and d_vocab must be explicitly set.
+        move_to_device (bool): Whether to move the model to the device specified in cfg.
+            device.
+        """
+        super().__init__()
+        self.__init_config__(config=cfg)
+        self.__init_tokenizer__(tokenizer=tokenizer)
+        self.__init_d_vocab__()
+        self.__init_embeddings__()
```
Comment (neelnanda-io) [inline easy_transformer/EasyTransformer.py:131]: Refactoring this out feels like overkill

File: easy_transformer/EasyTransformer.py [hunk 2]
```diff
@@ -94,21 +68,27 @@ def __init__(
...
 
-        self.blocks = nn.ModuleList(
-            [
-                TransformerBlock(self.cfg, block_index)
-                for block_index in range(self.cfg.n_layers)
-            ]
-        )
-
-        if self.cfg.normalization_type == "LN":
+    def __init__normalization__(self):
```
Comment (neelnanda-io) [inline easy_transformer/EasyTransformer.py:88]: Also overkill

Comment (neelnanda-io) [reply easy_transformer/EasyTransformer.py:88]: At least, called it final normalization or pre-unembed normalization

File: easy_transformer/EasyTransformer.py [hunk 3]
```diff
@@ -126,6 +106,39 @@ def __init__(
...
+        cfg Union[EasyTransformerConfig, Dict]: The config to use for the
+            model.
+        tokenizer (*optional): The tokenizer to use for the model. If not
+            provided, it is inferred from cfg.tokenizer_name or initialized to None.
+            If None, then the model cannot be passed strings, and d_vocab must be explicitly set.
+        move_to_device (bool): Whether to move the model to the device specified in cfg.
+            device.
+        """
+        super().__init__()
+        self.__init_config__(config=cfg)
```
Comment (neelnanda-io) [inline easy_transformer/EasyTransformer.py:128]: I overall like the vibe, but would refactor to return config, and do self.cfg = self.__init_config__(config=cfg). This is important bc then the user knows what everything is called.

File: easy_transformer/EasyTransformer.py [hunk 4]
```diff
@@ -248,6 +247,55 @@ def forward(
...
+        Note that loss is the standard "predict the next token" cross-entropy loss for GPT-2 style language models - if you want a custom loss function, the recommended behaviour is returning the logits and then applying your custom loss function.
+        """
+        tokens = self.__make_tokens_for_forward__(input=input, prepend_bos=prepend_bos)
+        pos_offset = self.__get_pos_offset_for_forward__(
+            tokens=tokens, past_kv_cache=past_kv_cache
+        )
+        (
+            residual,
+            shortformer_pos_embed,
+        ) = self.__get_residual_and_shortform_pos_embed__(
```
[the remark this answers is not in the document]
Comment (neelnanda-io) [reply easy_transformer/EasyTransformer.py:277]: Maybe rename to get embedding?
And the rename to residual just before the for loop

File: easy_transformer/evals.py [hunk 5]
```diff
@@ -3,12 +3,14 @@
 A file with some rough evals for models - I expect you to be likely better off using the HuggingFace evaluate library if you want to do anything properly, but this is here if you want it and want to eg cheaply and roughly compare models you've trained to baselines.
 """
 
+import einops
```
Comment (neelnanda-io) [inline easy_transformer/evals.py:6]: What's the algorithm for refactoring imports?

File: easy_transformer/hook_points.py [hunk 6]
```diff
@@ -1,149 +1,216 @@
...
+
+class FunctionalNameFilter(NamesFilter):
+    def __init__(self, f: Callable[[str], bool]):
+        self.f = f
+
+    def is_included(self, name: str) -> bool:
+        return self.f(name)
+
+
+class ConstantNameFilter(NamesFilter):
```
Comment (neelnanda-io) [inline easy_transformer/hook_points.py:24]: There should also be a name class for just inputting a single name (Which is actually the standard use case lol)

File: easy_transformer/hook_points.py [hunk 7]
```diff
@@ -1,149 +1,216 @@
...
 
 # %%
-# A helper class to get access to intermediate activations (inspired by Garcon)
-# It's a dummy module that is the identity function by default
-# I can wrap any intermediate activation in a HookPoint and get a convenient
-# way to add PyTorch hooks
+
+
+# TODO make sure this is used in all the right places
+class HookDirection(Enum):
```
Comment (neelnanda-io) [inline easy_transformer/hook_points.py:38]: I think this is way uglier? Or at least, people will now do add_hook(dir=1) which seems way worse than dir="fwd"? (Unless you expect them to import + use this enum)

File: easy_transformer/hook_points.py [hunk 8]
```diff
@@ -1,149 +1,216 @@
...
-        # Helper function that's mainly useful on EasyTransformer
-        # If it doesn't have this form, raises an error -
-        split_name = self.name.split(".")
-        return int(split_name[1])
+        """
+        Returns the layer index if the name has the form 'blocks.{layer}.{...}'
+        Helper function that's mainly useful on EasyTransformer
+        If it doesn't have this form, raises an error -
+        """
+        _, layer, _ = self.name.split(".")
```
Comment (neelnanda-io) [inline easy_transformer/hook_points.py:121]: This won't work, the list may be three or four elements long (also this should probably return None for hooks without a layer, rather than error lol)

File: easy_transformer/hook_points.py [hunk 9]
```diff
@@ -1,149 +1,216 @@
...
     WARNING: The main footgun with PyTorch hooking is that hooks are GLOBAL state. If you add a hook to the module, and then run it a bunch of times, the hooks persist. If you debug a broken hook and add the fixed version, the broken one is still there. To solve this, run_with_hooks will remove hooks at the start and end by default, and I recommend using reset_hooks liberally in your code.
 
     The main time this goes wrong is when you want to use backward hooks (to cache or intervene on gradients). In this case, you need to keep the hooks around as global state until you've run loss.backward() (and so need to disable the reset_hooks_end flag on run_with_hooks)
     """
 
-    def __init__(self, *args):
+    def __init__(self):
         super().__init__()
-        self.is_caching = False
+        self.is_caching = False  # TODO what's this used for?
```
Comment (neelnanda-io) [inline easy_transformer/hook_points.py:142]: Deprecated, can delete

File: easy_transformer/hook_points.py [hunk 10]
```diff
@@ -1,149 +1,216 @@
...
         # Build a dictionary mapping a module name to the module
-        self.mod_dict = {}
-        self.hook_dict = {}
+        self.name_to_hook = {}
         for name, module in self.named_modules():
             module.name = name
-            self.mod_dict[name] = module
+            self.name_to_module[
+                name
+            ] = module  # TODO why not just use self.named_modules()?
```
Comment (neelnanda-io) [inline easy_transformer/hook_points.py:157]: It's a generator not a dictionary, it's a headache

File: easy_transformer/hook_points.py [hunk 11]
```diff
@@ -1,149 +1,216 @@
...
+            # TODO is this really the best way to do this?
```
Comment (neelnanda-io) [inline easy_transformer/hook_points.py:158]: Yes. It's such a mess

Comment (neelnanda-io) [reply easy_transformer/hook_points.py:158]: I tried man. I tried.

Comment (smithjessk) [reply easy_transformer/hook_points.py:155]: what about a subclass?

Comment (smithjessk) [reply easy_transformer/hook_points.py]: (and then checking https://www.programiz.com/python-programming/methods/built-in/issubclass)

File: easy_transformer/hook_points.py (hunk 9 above)
Comment (smithjessk) [reply easy_transformer/hook_points.py]: wait, what about [this helpful warning? ](https://github.com/neelnanda-io/Easy-Transformer/blob/a540929a0ca34e321528f5e8b7716e30a0744ca9/easy_transformer/hook_points.py#L223)

File: easy_transformer/hook_points.py (hunk 8 above)
Comment (smithjessk) [reply easy_transformer/hook_points.py]: 6f67168

File: easy_transformer/evals.py (hunk 5 above)
Comment (smithjessk) [reply easy_transformer/evals.py:6]: [isort](https://github.com/PyCQA/isort)

> isort is a Python utility / library to sort imports alphabetically, and automatically separated into sections and by type.

File: easy_transformer/hook_points.py (hunk 6 above)
Comment (smithjessk) [reply easy_transformer/hook_points.py:24]: d17cff8

File: easy_transformer/hook_points.py (hunk 7 above)
Comment (smithjessk) [reply easy_transformer/hook_points.py]: bcbb687

File: easy_transformer/EasyTransformer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (smithjessk) [reply easy_transformer/EasyTransformer.py]: as in 0b7c598653b07f8ee22046120500c44f67eef9c4?

File: easy_transformer/hook_points.py (hunk 8 above)
Comment (smithjessk) [reply easy_transformer/hook_points.py]: how do you like this style? 23fbe80812c674d9ba522e3a707985f2e43015fa

Comment (smithjessk) [reply easy_transformer/hook_points.py]: (i will finish the rest if you approve!)

File: easy_transformer/EasyTransformer.py (hunk 3 above)
Comment (smithjessk) [reply easy_transformer/EasyTransformer.py]: how do you like this style? 23fbe80812c674d9ba522e3a707985f2e43015fa

i will do the rest in it if you're a fan

File: easy_transformer/EasyTransformer.py (hunk 2 above)
Comment (smithjessk) [reply easy_transformer/EasyTransformer.py:91]: paused on https://github.com/neelnanda-io/Easy-Transformer/pull/71#pullrequestreview-1181263509

File: easy_transformer/EasyTransformer.py (hunk 1 above)
Comment (smithjessk) [reply easy_transformer/EasyTransformer.py:134]: as discussed on slack (at least my read, it's not 100% clear to me), seems fine if we do the style `self.cfg = ...`

**Comment by smithjessk** [comment]
Note that `pytest` fails on a laptop without a GPU. (That's true in `main` as well). 

Here's a colab demonstrating that EasyTransformer tests still work: https://colab.research.google.com/drive/1MKJ6nkRTNWXqXQYK524Ojjxvn7NkuDfI?usp=sharing

File: easy_transformer/hook_points.py [hunk 12]
```diff
@@ -149,25 +235,15 @@ def run_with_hooks(
...
-                        hp.add_hook(hook, dir="fwd")
-        for name, hook in bwd_hooks:
-            if type(name) == str:
-                self.mod_dict[name].add_hook(hook, dir="bwd")
-            else:
-                # Otherwise, name is a Boolean function on names
-                for hook_name, hp in self.hook_dict:
-                    if name(hook_name):
-                        hp.add_hook(hook, dir="bwd")
+        if reset_hooks_start:
```
Comment (smithjessk) [inline easy_transformer/hook_points.py:240]: TBC, do we still want `reset_hooks_start`? @neelnanda-io

**Review by smithjessk — COMMENTED** [review]
did my self-review, LGTM pending your comments @neelnanda-io

**Comment by alan-cooney** [comment]
> Note that `pytest` fails on a laptop without a GPU. (That's true in `main` as well).
> 
> Here's a colab demonstrating that EasyTransformer tests still work: https://colab.research.google.com/drive/1MKJ6nkRTNWXqXQYK524Ojjxvn7NkuDfI?usp=sharing

This should be fixed with https://github.com/neelnanda-io/Easy-Transformer/pull/68

By the way, it looks like your `test_` files are manual tests? I've also renamed them in that PR as they will otherwise clash with test runners (e.g. Pytest assumes they are tests and tries to run them). In general if they are manual tests, I'd recommend converting them to Pytest tests so the CD can run them on each push.

**Comment by alan-cooney** [comment]
Closing as the codebase has changed a lot since this was written, and merging is likely now too complex. @smithjessk sorry about this - if you want to pull out some of the changes here into other PRs on the new codebase, I can review them quickly if that helps?
